README.md
## Exploration on Data Augmentation and Learning Rate Optimization on Detection of COVID-19 via Chest X-ray (CXR) Images

Use pretrained models on chest x-ray dataset in order to classify:

* Normal images
* Bacterial pneumonia
* COVID-19
* Non COVID-19 viral pneumonia

The pretrained models include:

* ResNet50
* SqueezeNet
* DenseNet121

## Table of contents
* [Usage](#usage)
* [Scripts](#scripts)
* [Changelog](#changelog)
* [Notebook](#notebook)
* [ReadMeArchieve](#readmearchieve)

## Usage
This project was implemented on Kaggle Notebook (GPU kernel). The best way to run the experiements is to upload the notebook in `/notebook/` directory to [kaggle notebook](), and link two following dataset:

* [Kaggle COVID-19 radiography dataset](https://www.kaggle.com/tawsifurrahman/covid19-radiography-database)
* [Kaggle Chest X-ray dataset](https://www.kaggle.com/paultimothymooney/chest-xray-pneumonia)

The notebook will perform the following steps end-to-end

* data preprocessing
* model training and inferencing
* metrics reporting

## Scripts
`/scripts/` folder includes:

* preprocess_data_set2.py
* ex2_pipe_1.py
* ex3_sched_1.py
* ex4_sched_1.py
* ex5.py

## Changelog
Please see [here](https://github.com/lilsummer/CS598_DL/blob/main/changelog.md) for the changelog

## Notebook
Notebook directory includes all the kaggle notebook used for the model pipeline. This includes

* `resnet-classifier` use pretrained resnet50
* `squeezenet-classifier` 
* `densenet-classifier` use densenet 121
* `1cycle-resnet50` use 1cycle learning rate optimization to fine-tune resnet50

## ReadMeArchieve
[Here](https://github.com/lilsummer/CS598_DL/blob/main/README_ARCHIVE.md) is a incomplete history of model training

changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [Unreleased]
## [1.2.1] - 2020-05-04
### Added
- Notebook for calculating PR-AUC, added to the table

## [1.2.0] - 2020-04-27
### Added
- Notebook for running reduceLRonPlateau 

## [1.1.4] - 2020-04-24
### Added
- Figures for the manuscript

## [1.1.3] - 2021-04-18
### Added
- New train-test setting for the model
### Changed
- Added new baseline by having more consistent augmentation between train and testing

## [1.1.2] - 2021-04-15
## Changed
- data augmentation: using Augmentor to increase the training set for COVID-19

## [1.1.1] - 2021-04-14
### Added
- pc_curve_0413_resnet_b.csv for baseline resent50 without rotation
- pc_curve_0414_resnet_distortion.csv for distortion resnet50 


## [1.1.0] - 2021-04-13
### Added
- pr_curve_0413_squeezenet10.csv pr curve on squeezenet
- pr_curve_0413_squeezenet10_b.csv without rotation 

## [1.0.4] - 2021-04-12
### Changed
- Added data processing so that no patient leakage in the bacteria dataset
### Added
- pr_curve_0412_a_no_removal.csv pr curve data with no patient removal in the bacteria dataset
- pr_curve_0412_b_removal.csv pr curve data with patient removal in the bacteria dataset


## [1.0.3] - 2021-04-11
### Changed
- Notebook change on saving the file for PR curve; the final image directory has to be removed in order to download the desired csv file.
### Added
- pr_curve_0410.csv corresponding to the model ran on 0410.

## [1.0.2] - 2021-04-04
### Added
- Randome sampling for training and testing sets
- Save statistics for PR curve
- Learning rate for new model


## [1.0.1] - 2021-04-03
### Added
- notebook (classifier_1_0_1) used to output PR-curve, history of loss, sensitivity and specificiy


## [1.0.0] - 2021-03-28
### Added
- a four-class classifier with bacteria pneumonia x-ray images
- model weights



scripts/ex5.py
## excludes data preprocessing, augmentation 
## excludes model evaluation

class CXRImgClf(nn.Module):
    def __init__(self):
        super(CXRImgClf, self).__init__()
        self.network= nn.Sequential(
            nn.Conv2d(3,8,3), 
            nn.ReLU(),
            nn.MaxPool2d(2,2),
            nn.Flatten(start_dim=1), #.view(-1,)
            nn.Linear(8*121*121,8*11*11),
            nn.Linear(8*11*11, 8*2*2),
            nn.Linear(8*2*2, 4)
        )
        
    def forward(self, x):
        return self.network(x)

def train_model(model, train_dataloader, n_epoch=n_epochs, optimizer=optimizer, criterion=criterion):
    import torch.optim as optim
    """
    :param model: A CNN model
    :param train_dataloader: the DataLoader of the training data
    :param n_epoch: number of epochs to train
    :return:
        model: trained model
    TODO:
        Within the loop, do the normal training procedures:
            pass the input through the model
            pass the output through loss_func to compute the loss (name the variable as *loss*)
            zero out currently accumulated gradient, use loss.basckward to backprop the gradients, then call optimizer.step
    """
    model.train() # prep model for training
    
    
    for epoch in range(n_epoch):
        curr_epoch_loss = []
        for data, target in train_dataloader:
            # your code here
            #raise NotImplementedError
            # pass the input through the model
            data = data.to(device)
            target = target.to(device)
            out_train = model(data)
            out_soft = torch.softmax(out_train, dim=1)
            
            # pass the output through loss_func
            loss = criterion(out_train, target)
            # zero out currently accumulated gradient, 
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            curr_epoch_loss.append(loss.cpu().data.numpy())
        print(f"Epoch {epoch}: curr_epoch_loss={np.mean(curr_epoch_loss)}")
    return model

scripts/ex2_pipe_1.py
## from processed data to model results
## this excldues preprossessing_data_set2.py

## imports
import torch
from torchvision import datasets, models, transforms
import torch.nn as nn
import torch.optim as optim

import pandas as pd
import numpy as np
import cv2
import matplotlib.pyplot as plt
import warnings
import random

import os
import itertools
import glob
import shutil

import Augmentor
from sklearn.metrics import precision_recall_curve


## augmentor
## Process covid directory first
p = Augmentor.Pipeline()
p.add_further_directory(new_source_directory='/kaggle/working/train_src/covid/', new_output_directory='/kaggle/working/train/covid/')
p.resize(width=244, height=244, probability=1)
p.crop_centre(probability=.1, percentage_area=.9)
p.resize(width=244, height=244, probability=1)
p.sample(3000)
## test dir 
p2 = Augmentor.Pipeline()
p2.add_further_directory(new_source_directory='/kaggle/working/test_src/covid/', new_output_directory='/kaggle/working/test/covid/')
p2.resize(width=244, height=244, probability=1)
p2.crop_centre(probability=.1, percentage_area=.9)
p2.resize(width=244, height=244, probability=1)

p2.process()

## valid dir
p3 = Augmentor.Pipeline()
p3.add_further_directory(new_source_directory='/kaggle/working/valid_src/covid/', new_output_directory='/kaggle/working/valid/covid/')
p3.resize(width=244, height=244, probability=1)
p3.crop_centre(probability=.1, percentage_area=.9)
p3.resize(width=244, height=244, probability=1)

p3.process()

## other directory are the same
for category in ['bacterial', 'viral', 'normal']:
    ## train
    p = Augmentor.Pipeline()
    p.add_further_directory(new_source_directory='/kaggle/working/train_src/' + category + '/', new_output_directory='/kaggle/working/train/' + category + '/')
    p.resize(width=244, height=244, probability=1)
    p.crop_centre(probability=.1, percentage_area=.9)
    p.resize(width=244, height=244, probability=1)

    p.process()
    
    ## test
    p2 = Augmentor.Pipeline()
    p2.add_further_directory(new_source_directory='/kaggle/working/test_src/' + category + '/', new_output_directory='/kaggle/working/test/' + \
                             category + '/')
    p2.resize(width=244, height=244, probability=1)
    p2.crop_centre(probability=.1, percentage_area=.9)
    p2.resize(width=244, height=244, probability=1)
    
    p2.process()
    
    ## valid
    p3 = Augmentor.Pipeline()
    p3.add_further_directory(new_source_directory='/kaggle/working/valid_src/' + category + '/', new_output_directory='/kaggle/working/valid/' + category+ '/')
    p3.resize(width=244, height=244, probability=1)
    p3.crop_centre(probability=.1, percentage_area=.9)
    p3.resize(width=244, height=244, probability=1)
    
    p3.process()

## remove all the train_src dir

files = glob.glob('/kaggle/working/train_src/*/*.png')
for f in files:
    os.remove(f)
files = glob.glob('/kaggle/working/test_src/*/*.png')
for f in files:
    os.remove(f)
files = glob.glob('/kaggle/working/valid_src/*/*.png')
for f in files:
    os.remove(f)
files = glob.glob('/kaggle/working/bacteria_merge/*.png')
for f in files:
    os.remove(f)

## normalize
normalizer = transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
data_transforms = {
    'train': transforms.Compose([
        transforms.ToTensor(),
        normalizer
    ]),
    'validation': transforms.Compose([
        transforms.ToTensor(),
        normalizer
    ])
}

## data loader
data_images = {
    'train': datasets.ImageFolder('/kaggle/working/train', data_transforms['train']),
    'validation': datasets.ImageFolder('/kaggle/working/valid', data_transforms['validation'])
}
dataloaders = {
    'train': torch.utils.data.DataLoader(data_images['train'], batch_size=32, shuffle=True, num_workers=0),
    'validation': torch.utils.data.DataLoader(data_images['validation'], batch_size=32,shuffle=False,num_workers=0)
}

testloaders = {
    'test': torch.utils.data.DataLoader(datasets.ImageFolder('/kaggle/working/test/', data_transforms['validation']), \
                                       batch_size=32, shuffle=False, num_workers=0)
}


## freeze the model
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
model = models.resnet50(pretrained=True).cuda()

for param in model.parameters():
    param.requires_grad = False

## replace last fc layer
model.fc = nn.Sequential(
    nn.Linear(2048, 64),
    nn.ReLU(inplace=True),
    nn.Linear(64, 4) # change the last dimension
).to(device)

criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.fc.parameters(), lr = .5e-3)

## training model function
def trained_model(model, criterion, optimizer, epochs):
    loss_history, acc_history = [],[]
    for epoch in range(epochs):
        
        ## report epoch
        print('Epoch:', str(epoch+1) + '/' + str(epochs))
        print('-'*10)

        ## confusion matrix
        conf_matrix = torch.zeros(4, 4)
        for phase in ['train', 'validation']:
            if phase == 'train':
                model.train() #this trains the model
            else:
                model.eval() #this evaluates the model

            running_loss, running_corrects = 0.0, 0 
            
            ## adding Y_test and Y_pred
            Y_test=[]
            Y_pred=[]
            Y_pred_prob = []
            for inputs, labels in dataloaders[phase]:
                inputs = inputs.to(device) #convert inputs to cpu or cuda
                labels = labels.to(device) #convert labels to cpu or cuda

                outputs = model(inputs) #outputs is inputs being fed to the model
                loss = criterion(outputs, labels) #outputs are fed into the model
                
                preds_soft = torch.softmax(outputs, dim=1)
                _, preds = torch.max(preds_soft, 1)
                running_loss += loss.item() * inputs.size(0) #loss multiplied by the first dimension of inputs
                running_corrects += torch.sum(preds == labels.data) #sum of all the correct predictions
                
                if phase == 'train':
                    optimizer.zero_grad() #sets gradients to zero
                    loss.backward() #computes sum of gradients
                    optimizer.step() #preforms an optimization step

                ## calculate sensitivity and specificity
                if phase == 'validation':
                    ## update confusion matrix
                    for t, p in zip(labels.data, preds):
                        conf_matrix[t, p] += 1
                
                ## if it is the last epoch record the results
                if phase=='validation' and epoch == epochs-1:
                    y_pred = preds.detach().cpu().numpy()
                    y_test = labels.data.detach().cpu().numpy()
                    y_pred_prob = preds_soft.detach().cpu().numpy()
                    
                    ## append to list
                    Y_pred_prob.append(y_pred_prob)
                    Y_test.append(y_test)
                    Y_pred.append(y_pred)
            
            ## calculate loss
            epoch_loss = running_loss / len(data_images[phase]) #this is the epoch loss
            epoch_accuracy = running_corrects.double() / len(data_images[phase]) #this is the epoch accuracy

            print(phase, ' loss:', epoch_loss, 'epoch_accuracy:', epoch_accuracy)
            
            loss_history.append(epoch_loss)
            acc_history.append(epoch_accuracy)
            
            if phase == 'validation':
                ## need to adjust the index based on the class index
                correct_covid_label = conf_matrix[1,1]
                total_covid_label = torch.sum(conf_matrix[1,:])
                correct_non_covid = conf_matrix[0,0] + conf_matrix[2,2] + conf_matrix[3,3]
                total_non_convid_label = torch.sum(conf_matrix[0,:]) + torch.sum(conf_matrix[2,:]) + torch.sum(conf_matrix[3,:])

                ## report sensitivity and specificity
                print('sensitivity: ', correct_covid_label/total_covid_label)
                print('specificity: ', correct_non_covid/total_non_convid_label)
    
    ## combine result
    Y_pred = np.concatenate(Y_pred, axis=0)
    Y_test = np.concatenate(Y_test, axis=0)
    Y_pred_prob =np.concatenate(Y_pred_prob, axis=0)
    
    return model, conf_matrix, Y_test, Y_pred, Y_pred_prob, loss_history, acc_history

## train model
model, mtrx, Y_test, Y_pred, Y_pred_prob, loss_his, acc_his = trained_model(model, criterion, optimizer, 20)

## eval function
def eval_model(model, dataloader):

    model.eval()
    Y_pred = []
    Y_test = []
    Y_pred_prob = []
    for data, target in dataloader:

        data = data.to(device)
        output = model(data)
        preds_soft = torch.softmax(output, dim=1)
        _, preds = torch.max(preds_soft, 1)
        y_pred = preds.detach().cpu().numpy()
        Y_pred.append(y_pred)
        
        y_test = target.data.detach().cpu().numpy()
        Y_test.append(y_test)
        
        y_pred_prob = preds_soft.detach().cpu().numpy()
        Y_pred_prob.append(y_pred_prob)
        #print(target)
        #print(Y_pred[0].shape)
    Y_pred = np.concatenate(Y_pred, axis=0)
    Y_test = np.concatenate(Y_test, axis=0)
    Y_pred_prob = np.concatenate(Y_pred_prob, axis=0)
    return Y_pred, Y_test, Y_pred_prob


## test on test set
Y_pred, Y_test, Y_pred_prob = eval_model(model, testloaders['test'])

## final conf matrix
conf_mtrx = np.zeros((4,4), dtype='int')
for t, p in zip(Y_test, Y_pred):
    conf_mtrx[t, p]+=1

## sensitivity & specificity
Y_test_COVID = [1 if i ==1 else 0 for i in Y_test]
Y_pred_COVID = [1 if i == 1 else 0 for i in Y_pred]

total_covid_label = np.sum(Y_test_COVID)
total_non_covid = len(Y_test_COVID) - total_covid_label
correct_covid_label = 0
correct_non_covid = 0
for i, v in enumerate(Y_test_COVID):
    if v==1 and Y_pred_COVID[i] == v:
        correct_covid_label +=1
    elif v==0 and Y_pred_COVID[i] ==v:
        correct_non_covid +=1
    else:
        pass

sensitivity = correct_covid_label/total_covid_label
specificity = correct_non_covid/total_non_covid

## pr curve
prec, recall, c = precision_recall_curve(Y_test_COVID , Y_pred_prob[:,1])

## save res
pd.DataFrame({'precision': prec, 'recall': recall}).to_csv('pc_curve_0420_crop_3.csv', index=False)

## remove necessary dataset
files = glob.glob('/kaggle/working/train/*/*.png')
for f in files:
    os.remove(f)
files = glob.glob('/kaggle/working/test/*/*.png')
for f in files:
    os.remove(f)
files = glob.glob('/kaggle/working/valid/*/*.png')
for f in files:
    os.remove(f)





scripts/ex3_sched_1.py
## from processed data to model results
## this excldues preprossessing_data_set2.py

## imports
import torch
from torchvision import datasets, models, transforms
import torch.nn as nn
import torch.optim as optim

import pandas as pd
import numpy as np
import cv2
import matplotlib.pyplot as plt
import warnings
import random

import os
import itertools
import glob
import shutil

import Augmentor
from sklearn.metrics import precision_recall_curve


## augmentor
## Process covid directory first
p = Augmentor.Pipeline()
p.add_further_directory(new_source_directory='/kaggle/working/train_src/covid/', new_output_directory='/kaggle/working/train/covid/')
p.resize(width=244, height=244, probability=1)
p.crop_centre(probability=.1, percentage_area=.9)
p.resize(width=244, height=244, probability=1)
p.sample(3000)
## test dir 
p2 = Augmentor.Pipeline()
p2.add_further_directory(new_source_directory='/kaggle/working/test_src/covid/', new_output_directory='/kaggle/working/test/covid/')
p2.resize(width=244, height=244, probability=1)
p2.crop_centre(probability=.1, percentage_area=.9)
p2.resize(width=244, height=244, probability=1)

p2.process()

## valid dir
p3 = Augmentor.Pipeline()
p3.add_further_directory(new_source_directory='/kaggle/working/valid_src/covid/', new_output_directory='/kaggle/working/valid/covid/')
p3.resize(width=244, height=244, probability=1)
p3.crop_centre(probability=.1, percentage_area=.9)
p3.resize(width=244, height=244, probability=1)

p3.process()

## other directory are the same
for category in ['bacterial', 'viral', 'normal']:
    ## train
    p = Augmentor.Pipeline()
    p.add_further_directory(new_source_directory='/kaggle/working/train_src/' + category + '/', new_output_directory='/kaggle/working/train/' + category + '/')
    p.resize(width=244, height=244, probability=1)
    p.crop_centre(probability=.1, percentage_area=.9)
    p.resize(width=244, height=244, probability=1)

    p.process()
    
    ## test
    p2 = Augmentor.Pipeline()
    p2.add_further_directory(new_source_directory='/kaggle/working/test_src/' + category + '/', new_output_directory='/kaggle/working/test/' + \
                             category + '/')
    p2.resize(width=244, height=244, probability=1)
    p2.crop_centre(probability=.1, percentage_area=.9)
    p2.resize(width=244, height=244, probability=1)
    
    p2.process()
    
    ## valid
    p3 = Augmentor.Pipeline()
    p3.add_further_directory(new_source_directory='/kaggle/working/valid_src/' + category + '/', new_output_directory='/kaggle/working/valid/' + category+ '/')
    p3.resize(width=244, height=244, probability=1)
    p3.crop_centre(probability=.1, percentage_area=.9)
    p3.resize(width=244, height=244, probability=1)
    
    p3.process()

## remove all the train_src dir

files = glob.glob('/kaggle/working/train_src/*/*.png')
for f in files:
    os.remove(f)
files = glob.glob('/kaggle/working/test_src/*/*.png')
for f in files:
    os.remove(f)
files = glob.glob('/kaggle/working/valid_src/*/*.png')
for f in files:
    os.remove(f)
files = glob.glob('/kaggle/working/bacteria_merge/*.png')
for f in files:
    os.remove(f)

## normalize
normalizer = transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
data_transforms = {
    'train': transforms.Compose([
        transforms.ToTensor(),
        normalizer
    ]),
    'validation': transforms.Compose([
        transforms.ToTensor(),
        normalizer
    ])
}

## data loader
data_images = {
    'train': datasets.ImageFolder('/kaggle/working/train', data_transforms['train']),
    'validation': datasets.ImageFolder('/kaggle/working/valid', data_transforms['validation'])
}
dataloaders = {
    'train': torch.utils.data.DataLoader(data_images['train'], batch_size=32, shuffle=True, num_workers=0),
    'validation': torch.utils.data.DataLoader(data_images['validation'], batch_size=32,shuffle=False,num_workers=0)
}

testloaders = {
    'test': torch.utils.data.DataLoader(datasets.ImageFolder('/kaggle/working/test/', data_transforms['validation']), \
                                       batch_size=32, shuffle=False, num_workers=0)
}


## add device
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')

## add loss
criterion = nn.CrossEntropyLoss()

## baseline class
class ImageClassificationBase(nn.Module):
    def training_step(self, batch):
        images, labels = batch 
        out = self(images)                  # Generate predictions
        loss =F.cross_entropy(out, labels) # Calculate loss
        return loss
    
    def validation_step(self, batch):
        images, labels = batch 
        out = self(images)                    # Generate predictions
        loss = F.cross_entropy(out, labels)   # Calculate loss
        #acc = accuracy(out, labels)           # Calculate accuracy
        return {'val_loss': loss.detach()}
        
    def validation_epoch_end(self, outputs):
        batch_losses = [x['val_loss'] for x in outputs]
        epoch_loss = torch.stack(batch_losses).mean()   # Combine losses
        #batch_accs = [x['val_acc'] for x in outputs]
        #epoch_acc = torch.stack(batch_accs).mean()      # Combine accuracies
        return {'val_loss': epoch_loss.item()}
    
    def epoch_end(self, epoch, result):
        print("Epoch [{}], last_lr: {:.5f}, train_loss: {:.4f}, val_loss: {:.4f}".format(
            epoch, result['lrs'][-1], result['train_loss'], result['val_loss']))

## resnet extend on base class
class XrayRES(ImageClassificationBase):
    def __init__(self, num_classes):
        super().__init__()
        self.network = models.resnet50()
        ## last fully-connected
        num_ftrs = self.network.fc.in_features
        self.network.fc = \
        nn.Sequential(nn.Linear(2048, 64),
                      nn.ReLU(inplace=True),
                      nn.Linear(64, 4))
        #num_ftrs = self.network.classifier[-1].in_features
        #self.network.classifier[-1] = nn.Linear(in_features=num_ftrs, out_features=num_classes)
    def forward(self, xb):
        return torch.softmax(self.network(xb), dim=1)
    
    def freeze(self):
        # to freeze the residual layers
        for param in self.network.parameters():
            param.require_grad = False
            
    def unfreeze(self):
        for param in self.network.parameters():
            param.require_grad = True

## auxillary functions
def get_default_device():
    """Pick GPU if available, else CPU"""
    if torch.cuda.is_available():
        return torch.device('cuda')
    else:
        return torch.device('cpu')
    
def to_device(data, device):
    """Move tensor(s) to chosen device"""
    if isinstance(data, (list,tuple)):
        return [to_device(x, device) for x in data]
    return data.to(device, non_blocking=True)

class DeviceDataLoader():
    """Wrap a dataloader to move data to a device"""
    def __init__(self, dl, device):
        self.dl = dl
        self.device = device
        
    def __iter__(self):
        """Yield a batch of data after moving it to device"""
        for b in self.dl: 
            yield to_device(b, self.device)

    def __len__(self):
        """Number of batches"""
        return len(self.dl)
model = to_device(XrayRES(4), device)

## more functions
@torch.no_grad()
def evaluate(model, val_loader):
    model.eval()
    outputs = [model.validation_step(batch) for batch in val_loader]
    return model.validation_epoch_end(outputs)

def get_lr(optimizer):
    for param_group in optimizer.param_groups:
        return param_group['lr']

def fit_one_cycle(epochs, max_lr, model, train_loader, val_loader, 
                  weight_decay=0, grad_clip=None, opt_func=torch.optim.Adam):
    torch.cuda.empty_cache()
    history = []
    
    # Set up cutom optimizer with weight decay
    optimizer = opt_func(model.parameters(), max_lr, weight_decay=weight_decay)
    # Set up one-cycle learning rate scheduler
    sched = torch.optim.lr_scheduler.OneCycleLR(optimizer, max_lr, epochs=epochs, 
                                                steps_per_epoch=len(train_loader))
    
    for epoch in range(epochs):
        # Training Phase 
        model.train()
        train_losses = []
        lrs = []
        for batch in train_loader:
            loss = model.training_step(batch)
            train_losses.append(loss)
            loss.backward()
            
            # Gradient clipping
            if grad_clip: 
                nn.utils.clip_grad_value_(model.parameters(), grad_clip)
            
            optimizer.step()
            optimizer.zero_grad()
            
            # Record & update learning rate
            lrs.append(get_lr(optimizer))
            sched.step()
        
        # Validation phase
        result = evaluate(model, val_loader)
        result['train_loss'] = torch.stack(train_losses).mean().item()
        result['lrs'] = lrs
        model.epoch_end(epoch, result)
        history.append(result)
    return history

## eval model first
val_dataloader = DeviceDataLoader(dataloaders['validation'], device)
train_dataloader = DeviceDataLoader(dataloaders['train'], device)
history = [evaluate(model, val_dataloader)]

## freeze
model.freeze()

## train 15 epochs
epochs = 15
max_lr = 1e-3
grad_clip = 0.1
weight_decay = 1e-4
opt_func = torch.optim.Adam

##%%time
history += fit_one_cycle(epochs, max_lr, model, train_dataloader, val_dataloader, 
                         grad_clip=grad_clip, 
                         weight_decay=weight_decay, 
                         opt_func=opt_func)

## train 8 epochs
epochs = 8
max_lr = 5e-4
grad_clip = 0.1
weight_decay = 1e-4
opt_func = torch.optim.Adam
history += fit_one_cycle(epochs, max_lr, model, train_dataloader, val_dataloader, 
                         grad_clip=grad_clip, 
                         weight_decay=weight_decay, 
                         opt_func=opt_func)


## train 5 epochs
epochs = 5
max_lr = 5e-4
grad_clip = 0.1
weight_decay = 1e-4
opt_func = torch.optim.Adam

history += fit_one_cycle(epochs, max_lr, model, train_dataloader, val_dataloader, 
                         grad_clip=grad_clip, 
                         weight_decay=weight_decay, 
                         opt_func=opt_func)

train_loss = [i['train_loss'] for i in history[1:]]
val_loss = [i['val_loss'] for i in history[1:]]
lr = [i['lrs'] for i in history[1:]]

## plot lr
def plot_lrs(history):
    lrs = np.concatenate([x.get('lrs', []) for x in history])
    plt.figure(figsize=(8, 5))
    plt.plot(lrs)
    plt.xlabel('Batch no.')
    plt.ylabel('Learning rate')
    plt.title('Learning Rate vs. Batch no.');


plot_lrs(history)
plt.tight_layout()
plt.savefig('sched_3_2.pdf')

## eval function
def eval_model(model, dataloader):

    model.eval()
    Y_pred = []
    Y_test = []
    Y_pred_prob = []
    for data, target in dataloader:

        data = data.to(device)
        output = model(data)
        preds_soft = torch.softmax(output, dim=1)
        _, preds = torch.max(preds_soft, 1)
        y_pred = preds.detach().cpu().numpy()
        Y_pred.append(y_pred)
        
        y_test = target.data.detach().cpu().numpy()
        Y_test.append(y_test)
        
        y_pred_prob = preds_soft.detach().cpu().numpy()
        Y_pred_prob.append(y_pred_prob)
        #print(target)
        #print(Y_pred[0].shape)
    Y_pred = np.concatenate(Y_pred, axis=0)
    Y_test = np.concatenate(Y_test, axis=0)
    Y_pred_prob = np.concatenate(Y_pred_prob, axis=0)
    return Y_pred, Y_test, Y_pred_prob


## test on test set
Y_pred, Y_test, Y_pred_prob = eval_model(model, testloaders['test'])

## final conf matrix
conf_mtrx = np.zeros((4,4), dtype='int')
for t, p in zip(Y_test, Y_pred):
    conf_mtrx[t, p]+=1

## sensitivity & specificity
Y_test_COVID = [1 if i ==1 else 0 for i in Y_test]
Y_pred_COVID = [1 if i == 1 else 0 for i in Y_pred]

total_covid_label = np.sum(Y_test_COVID)
total_non_covid = len(Y_test_COVID) - total_covid_label
correct_covid_label = 0
correct_non_covid = 0
for i, v in enumerate(Y_test_COVID):
    if v==1 and Y_pred_COVID[i] == v:
        correct_covid_label +=1
    elif v==0 and Y_pred_COVID[i] ==v:
        correct_non_covid +=1
    else:
        pass

sensitivity = correct_covid_label/total_covid_label
specificity = correct_non_covid/total_non_covid

## pr curve
prec, recall, c = precision_recall_curve(Y_test_COVID , Y_pred_prob[:,1])

## save res
pd.DataFrame({'precision': prec, 'recall': recall}).to_csv('pc_curve_0420_crop_3.csv', index=False)

## remove necessary dataset
files = glob.glob('/kaggle/working/train/*/*.png')
for f in files:
    os.remove(f)
files = glob.glob('/kaggle/working/test/*/*.png')
for f in files:
    os.remove(f)
files = glob.glob('/kaggle/working/valid/*/*.png')
for f in files:
    os.remove(f)
